x/incentive/keeper/reward_tracker.go

```diff
@@ -0,0 +1,295 @@
+// the creation of a new period, which initializes the FP, creates
+// historical reward tracker, withdraw the BTC delegation rewards to gauge
+// and initializes a new delegation with the just ended period.
+func (k Keeper) BtcDelegationActivated(ctx context.Context, fp, del sdk.AccAddress, sat uint64) error {
+	amtSat := sdkmath.NewIntFromUint64(sat)
+	return k.btcDelegationModifiedWithPreInitDel(ctx, fp, del, func(ctx context.Context, fp, del sdk.AccAddress) error {
+		return k.addDelegationSat(ctx, fp, del, amtSat)
+	})
+}
+
+// BtcDelegationUnbonded it modifies the total amount of satoshi staked
+// for the delegation (fp, del) and for the finality provider by subtracting.
+// Since it modifies the active amount it triggers the increment of fp period,
+// creationg of new historical reward, withdraw of rewards to gauge
+// and initialization of a new delegation.
+// It errors out if the unbond amount is higher than the total amount staked.
+func (k Keeper) BtcDelegationUnbonded(ctx context.Context, fp, del sdk.AccAddress, sat uint64) error {
+	amtSat := sdkmath.NewIntFromUint64(sat)
+	return k.btcDelegationModifiedWithPreInitDel(ctx, fp, del, func(ctx context.Context, fp, del sdk.AccAddress) error {
+		return k.subDelegationSat(ctx, fp, del, amtSat)
+	})
+}
+
+// FpSlashed a slashed finality provider should withdraw all the rewards
+// available to it, by iterating over all the delegations for this FP
+// and sending to the gauge. After the rewards are removed, it should
+// delete every rewards tracker value in the store related to this slashed
+// finality provider.
+func (k Keeper) FpSlashed(ctx context.Context, fp sdk.AccAddress) error {
+	// finalize the period to get a new history with the current rewards available
+	endedPeriod, err := k.IncrementFinalityProviderPeriod(ctx, fp)
+	if err != nil {
+		return err
+	}
+
+	// remove all the rewards available from the just ended period
+	keysBtcDelRwdTracker := make([][]byte, 0)
+	if err := k.IterateBTCDelegationRewardsTracker(ctx, fp, func(fp, del sdk.AccAddress) error {
+		keysBtcDelRwdTracker = append(keysBtcDelRwdTracker, del.Bytes())
+		return k.CalculateBTCDelegationRewardsAndSendToGauge(ctx, fp, del, endedPeriod)
+	}); err != nil {
+		return err
+	}
+
+	// delete all reward tracker that correlates with the slashed finality provider.
+	k.deleteKeysFromBTCDelegationRewardsTracker(ctx, fp, keysBtcDelRwdTracker)
+	k.deleteAllFromFinalityProviderRwd(ctx, fp)
+	return nil
+}
+
+// sendAllBtcRewardsToGauge iterates over all the finality providers associated
+// with the delegator and withdraw the rewards available to the gauge.
+// This creates new periods for each delegation and finality provider.
+func (k Keeper) sendAllBtcRewardsToGauge(ctx context.Context, del sdk.AccAddress) error {
+	return k.iterBtcDelegationsByDelegator(ctx, del, func(del, fp sdk.AccAddress) error {
+		return k.btcDelegationModified(ctx, fp, del)
+	})
+}
```

Comment:
It iterates over all the delegations from one address and acts as if the btc delegation was modified (which was, since BTC rewards are being withdraw) it just doesn't modifies the amount staked